### PR TITLE
Merge @zostay mutator work

### PR DESCRIPTION
**Pull request by jonathanstowe**
This is to cover off #27 fully rebased and all tests working.

I wouldn't recommend merging this unless it's tested with some modules that use it (i.e. things that can't be done in the unit tests.)

### CODE REVIEW

**Comment by jonathanstowe** [comment]
It's still not passing on Appveyor but I'm not sure what that's about.

**Review by Scimon — APPROVED** [review]
So there's a *lot* here but it all looks sane (there were a couple of methods it took me a moment to twig (positional vs named basically)). 
The test suite looks complete so I would err on the side of merging it. If there are bugs they will be edge cases we can handle in the future.

File: appveyor.yml [hunk 1]
```diff
@@ -17,6 +17,7 @@ install:
   - git clone https://github.com/tadzik/rakudobrew %USERPROFILE%\rakudobrew
   - SET PATH=%USERPROFILE%\rakudobrew\bin;%PATH%
   - rakudobrew build moar %TEST_MOAR%
+  - SET PATH=%USERPROFILE%\rakudobrew\versions\moar-%TEST_MOAR%\install\bin:%PATH%
```
Comment (Scimon) [inline appveyor.yml]: This is the issues with appveyor failing I think should be 

SET PATH=%USERPROFILE%\rakudobrew\versions\moar-%TEST_MOAR%\install\bin;%PATH%

**Review by Scimon — CHANGES_REQUESTED** [review]
The rest of my review stands but I think this is why Appveyor is faling. Windows path seperator.

File: appveyor.yml (hunk 1 above)
Comment (jonathanstowe) [reply appveyor.yml]: Hahah yeah cheers.

Comment (jonathanstowe) [reply appveyor.yml]: Well that definitely moved on the appveyor to a different fail in zef.

Comment (Scimon) [reply appveyor.yml]: Yeah. Might it be the Moar version?

**Comment by Scimon** [comment]
I think the test-moar: '' doesn't work because it tries to set the path to moar-/bin to get perl6.
I think you need to set the version. (Also the other is runing on 2017-05. I'm not suprised if zef won't install.

**Review by Scimon — APPROVED** [review]

**Comment by jonathanstowe** [comment]
Eek is there any way to set the version to that which rakudobrew  just built?

**Comment by jonathanstowe** [comment]
Okay on the grounds that no-one has touched the appveyor config since it was added and it has been failing consistently I'm going to remove it.  If anyone feels strongly enough they can make it work again.

**Comment by jonathanstowe** [comment]
I'm going to merge this manually to up version, Changes, etc easier.

**Comment by jonathanstowe** [comment]
Merged with abe8c9bb65947760cb656c6c154f466cd87f6e57  as v0.3.0
